certbot/certbot/_internal/renewal.py

```diff
@@ -295,12 +295,13 @@ def should_renew(config, lineage):
         logger.debug("Auto-renewal forced with --force-renewal...")
         return True
     if lineage.should_autorenew():
-        logger.info("Cert is due for renewal, auto-renewing...")
+        logger.info("Certificate is due for renewal, auto-renewing...")
         return True
     if config.dry_run:
-        logger.info("Cert not due for renewal, but simulating renewal for dry run")
+        logger.info("Certificate not due for renewal, but simulating renewal for dry run")
         return True
-    logger.info("Cert not yet due for renewal")
+    display_util.notify("Certificate not yet due for renewal")
+    logger.debug("Certificate not yet due for renewal")
```

Comment:
This one ends up in the debug log twice:

    2021-05-13 16:24:48,726:DEBUG:certbot.display.util:Notifying user: Certificate not yet due for renewal
    2021-05-13 16:24:48,726:DEBUG:certbot._internal.renewal:Certificate not yet due for renewal

In terms of a code pattern across Certbot, I've been assuming that sending that first `DEBUG` log line is part of `notify`'s semantics. So I've been rolling print/log into one `notify` call, basically.

If that's no good, I'm happy to adopt this pattern instead.

Is it possible that the reason you had trouble with a unit test here is that `getUtility` is mocked, which masks the subsequent call to `logger.debug`?

Reply:
Ah, I forgot it did that, you're absolutely correct. And that would certainly explain the test failure, which is a bit annoying.